### PR TITLE
flyingpio & simulator: receiver and ADC support

### PR DESCRIPTION
**Pull request by mlyle**
ADC functionality tested through battery module.  PIOS_RCVR failure cases tested-- correct timeout, missing driver stuff appears, but no actual receiver tested yet.

### CODE REVIEW

**Comment by mlyle** [comment]
0x2f8 data + bss.

Total stacks harder-- main 136, plus 352 for process_pio_message.  worst obvious IRQ handler is 88.  But USART1_IRQHandler (8) calls a function pointer... COM_RxInCallback takes 88, COM_TxOutCallback takes 104.

so by my reading thats 136+352+8+104, or 0x258.  

Total data+bss+stack seems to be 0x550.  Leaves 2736 bytes left for heap..

**Comment by mlyle** [comment]
IMO ready for review, even though it has not seen comprehensive test.

**Comment by mlyle** [comment]
Force-pushed.  I just tested this with sbus.  It works, but I had to add errata: silkscreen for 'rcvr' and 'serial' are swapped.

**Comment by mlyle** [comment]
jenkins, build artifacts please

File: flight/PiOS/Common/pios_hal.c [hunk 1]
```diff
@@ -761,7 +770,6 @@ void PIOS_HAL_ConfigurePort(HwSharedPortTypesOptions port_type,
 				PIOS_Assert(0);
 			}
 			PIOS_HAL_SetReceiver(MANUALCONTROLSETTINGS_CHANNELGROUPS_SBUS, sbus_rcvr_id);
-			pios_rcvr_group_map[MANUALCONTROLSETTINGS_CHANNELGROUPS_SBUS] = sbus_rcvr_id;
```
Comment (tracernz) [inline flight/PiOS/Common/pios_hal.c:764]: Dunno why this was here. I think we've discussed it before but never actually got round to removing it. :+1:

File: flight/PiOS/Common/pios_rcvr.c [hunk 2]
```diff
@@ -34,23 +33,6 @@ static struct pios_rcvr_dev * PIOS_RCVR_alloc(void)
...
-  if (pios_rcvr_num_devs >= PIOS_RCVR_MAX_DEVS) {
-    return (NULL);
-  }
-
-  rcvr_dev = &pios_rcvr_devs[pios_rcvr_num_devs++];
-  rcvr_dev->magic = PIOS_RCVR_DEV_MAGIC;
-
-  return (rcvr_dev);
-}
-#endif
```
Comment (tracernz) [inline flight/PiOS/Common/pios_rcvr.c:53]: We are discarding the ability to use receivers without RTOS? Or did you right a heap impl. that just uses a big static array? I forget.

File: flight/PiOS/posix/pios_flyingpio.c [hunk 3]
```diff
@@ -50,16 +51,38 @@ enum pios_flyingpio_dev_magic {
...
+
 /**
  * @brief The device state struct
  */
 struct pios_flyingpio_dev {
-	enum pios_flyingpio_dev_magic magic;              /**< Magic bytes to validate the struct contents */
-	uint32_t spi_id;                            /**< Handle to the communication driver */
-	uint32_t spi_slave;                         /**< The slave number (SPI) */
+	enum pios_flyingpio_dev_magic magic;    /**< Magic bytes to validate the struct contents */
+	uint32_t spi_id;		/**< Handle to the communication driver */
```
Comment (tracernz) [inline flight/PiOS/posix/pios_flyingpio.c]: Should use spaces for this.

File: flight/PiOS/posix/pios_flyingpio.c [hunk 4]
```diff
@@ -131,6 +154,11 @@ int32_t PIOS_FLYINGPIO_SPI_Init(pios_flyingpio_dev_t *dev, uint32_t spi_id, uint
 	fpio_dev->spi_id = spi_id;
 	fpio_dev->spi_slave = slave_idx;
 
+	for (int i=0; i<FPPROTO_MAX_RCCHANS; i++) {
```
Comment (tracernz) [inline flight/PiOS/posix/pios_flyingpio.c]: Missing spaces :stuck_out_tongue:

**Comment by tracernz** [comment]
Looks pretty good. :+1:

File: flight/PiOS/Common/pios_rcvr.c (hunk 2 above)
Comment (mlyle) [inline flight/PiOS/Common/pios_rcvr.c:53]: There already was a heap implementation that uses an array from SHEAP (end of BSS) to end of memory...  So this stuff had no purpose except to waste memory.

File: flight/PiOS/Common/pios_hal.c (hunk 1 above)
Comment (mlyle) [inline flight/PiOS/Common/pios_hal.c:764]: Yup.  now it'd actually be fatal.

File: flight/PiOS/Common/pios_rcvr.c (hunk 2 above)
Comment (mlyle) [inline flight/PiOS/Common/pios_rcvr.c:53]: (note that I _am_ using receivers without RTOS)

**Comment by mlyle** [comment]
Comments addressed-- thanks for review!

File: flight/PiOS/STM32F0xx/pios_delay.c [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+}
+
+float PIOS_RTC_MsPerTick()
+{
+	return 1000 / ((float) SYSTICK_HZ) * RTC_DIVIDER;
+}
+
+bool PIOS_RTC_RegisterTickCallback(void (*fn)(uintptr_t id), uintptr_t data)
+{
+	for (int i=0; i < PIOS_RTC_MAX_CALLBACKS; i++) {
```
Comment (tracernz) [inline flight/PiOS/STM32F0xx/pios_delay.c]: I hate to be annoying but still missing spaces around = operator. :/

File: flight/PiOS/STM32F0xx/pios_delay.c [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static void PIOS_RTC_Tick()
+{
+	for (int i=0; i < PIOS_RTC_MAX_CALLBACKS; i++) {
```
Comment (tracernz) [inline flight/PiOS/STM32F0xx/pios_delay.c]: And here.

File: flight/PiOS/posix/pios_flyingpio.c [hunk 7]
```diff
@@ -131,6 +154,11 @@ int32_t PIOS_FLYINGPIO_SPI_Init(pios_flyingpio_dev_t *dev, uint32_t spi_id, uint
 	fpio_dev->spi_id = spi_id;
 	fpio_dev->spi_slave = slave_idx;
 
+	for (int i=0; i < FPPROTO_MAX_RCCHANS; i++) {
```
Comment (tracernz) [inline flight/PiOS/posix/pios_flyingpio.c]: and here

File: flight/PiOS/posix/pios_flyingpio.c [hunk 8]
```diff
@@ -163,20 +191,44 @@ static int PIOS_FLYINGPIO_SendCmd(struct flyingpi_msg *msg) {
 	/* Handle response data-- after releasing SPI */
 	if (!ret) {
 		if (!flyingpi_calc_crc(&resp, false, NULL)) {
-			printf("fpio: VERYBAD: Bad CRC on response data %d\n",
+			printf("fpio: VERYBAD: Bad CRC on response data %d\n\t",
 				fpio_dev->msg_num);
+
+			uint8_t *msg_data = (void *) &resp;
+
+			for (int i=0; i<16; i++) {
```
Comment (tracernz) [inline flight/PiOS/posix/pios_flyingpio.c]: = and < here

File: flight/PiOS/posix/pios_flyingpio.c [hunk 9]
```diff
@@ -163,20 +191,44 @@ static int PIOS_FLYINGPIO_SendCmd(struct flyingpi_msg *msg) {
...
 			if (data->valid_messages_recvd != fpio_dev->msg_num) {
-				printf("fpio: BAD: sequence number mismatch\n");
+				printf("fpio: BAD: sequence number mismatch %d vs %d\n", data->valid_messages_recvd, fpio_dev->msg_num);
 
 				// Signal that caller should reconfig.
 				ret = -1;
 
 				fpio_dev->msg_num = data->valid_messages_recvd;
 			} else {
+				for (int i=0; i < FPPROTO_MAX_RCCHANS; i++) {
```
Comment (tracernz) [inline flight/PiOS/posix/pios_flyingpio.c]: #

File: flight/PiOS/posix/pios_flyingpio.c [hunk 10]
```diff
@@ -163,20 +191,44 @@ static int PIOS_FLYINGPIO_SendCmd(struct flyingpi_msg *msg) {
...
+					fpio_dev->rcvr_value[i] = data->chan_data[i];
+				}
+
+				for (int i=0; i<FPPROTO_MAX_ADCCHANS; i++) {
```
Comment (tracernz) [inline flight/PiOS/posix/pios_flyingpio.c]: = and <

**Pull request merged**